driver-core/src/main/com/mongodb/connection/CommandMessage.java

```diff
@@ -17,36 +17,180 @@
+
+            addDocument(getCommandToEncode(), bsonOutput, commandFieldNameValidator, getExtraElements(sessionContext));
+
+            if (payload != null) {
+                bsonOutput.writeByte(1);          // payload type
+                int payloadPosition = bsonOutput.getPosition();
+                bsonOutput.writeInt32(0);         // size
+                bsonOutput.writeCString(payload.getPayloadName());
+                writePayload(new BsonBinaryWriter(bsonOutput, payloadFieldNameValidator), bsonOutput, getSettings(),
+                        commandStartPosition, payload);
+
+                int payloadLength = bsonOutput.getPosition() - payloadPosition;
+                bsonOutput.writeInt32(payloadPosition, payloadLength);
+            }
+
+            // Write the flag bits
+            bsonOutput.writeInt32(flagPosition, getFlagBits());
+        } else {
+            bsonOutput.writeInt32(0);
+            bsonOutput.writeCString(namespace.getFullName());
+            bsonOutput.writeInt32(0);
+            bsonOutput.writeInt32(-1);
+
+            commandStartPosition = bsonOutput.getPosition();
+
+            if (payload == null) {
+                addDocument(getCommandToEncode(), bsonOutput, commandFieldNameValidator, null);
+            } else {
+                addDocumentWithPayload(bsonOutput);
+            }
+        }
+        return new EncodingMetadata(null, commandStartPosition);
     }
 
-    protected static OpCode getOpCode(final MessageSettings settings) {
-        return useOpMsg(settings) ? OpCode.OP_MSG : OpCode.OP_QUERY;
+    private FieldNameValidator getPayloadArrayFieldNameValidator() {
+        Map<String, FieldNameValidator> rootMap = new HashMap<String, FieldNameValidator>();
+        rootMap.put(payload.getPayloadName(), payloadFieldNameValidator);
+        return new MappedFieldNameValidator(commandFieldNameValidator, rootMap);
     }
 
-    protected static boolean useOpMsg(final MessageSettings settings) {
-        return isServerVersionAtLeastThreeDotSix(settings);
+    private void addDocumentWithPayload(final BsonOutput bsonOutput) {
+        BsonWriter writer = new BsonBinaryWriter(bsonOutput, getPayloadArrayFieldNameValidator());
+        if (payload != null) {
+            writer =  new SplittablePayloadBsonWriter(writer, bsonOutput, getSettings(), payload);
+        }
+        BsonDocument commandToEncode = getCommandToEncode();
+        getCodec(commandToEncode).encode(writer, commandToEncode, EncoderContext.builder().build());
+    }
+
+    private int getFlagBits() {
+        if (calculateIsResponseExpected()) {
+            return 0;
+        } else {
+            return 1 << 1;
+        }
+    }
+
+    private boolean calculateIsResponseExpected() {
+        // If there is another message in the payload require that the response is acknowledged
+        if (useOpMsg() && payload != null && payload.hasAnotherSplit()) {
+            return true;
```

Comment:
I confirmed in the debugger than no functional tests in driver-core hit line 152.  

There should be tests that force payload splitting for both acknowledged and unacknowledged writes that confirm that all the writes go through.

Reply:
Updated.